File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # null case
        if prices is None:
            return 0
        
        n = len(prices)
        max_profit = 0
        
        # for i in range(n-1):
        #     # valley condition
        #     if prices[i] > prices[i+1]:
        #         continue
        #     profit = prices[i+1] - prices[i]
        #     max_profit += profit
        # return max_profit
        
        peak = prices[0]
        valley = prices[0]
        i=0
        
        while i<n-1:
            # valley condition
            if i<n-1 and prices[i] >= prices[i+1]:
                i+=1
            valley = prices[i]
            if i<n-1 and prices[i] <= prices[i+1]:
                i+=1
            peak = prices[i]
            max_profit += peak - valley
        return max_profit
            

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if(n == 0) return 0;
        int maxProfit = 0;
        int peak = prices[0];
        int valley = prices[0];
        int i = 0;
        
        while(i<n-1){
            if((i<n-1) && (prices[i] >= prices[i+1])) i++;
            valley = prices[i];
            if((i<n-1) && (prices[i] <= prices[i+1])) i++;
            peak = prices[i];
            maxProfit += peak-valley;
        }
        return maxProfit;
    }
}